### DISCUSSION TITLE
Is any specification there?

### DISCUSSION THREAD

**Discussion by cathaysia** [pallets/jinja#2018] [Q&A] [answered]
Hello, I am writing a tree-sitter parser for jinja2. Currently, I write this parser rely on https://jinja.palletsprojects.com/en/3.1.x/templates/, but did jinja2 has any specification?

ref:

- https://github.com/cathaysia/tree-sitter-jinja2

**Comment by davidism** [comment, marked as answer]
There is an open issue about writing a spec for the grammar and runtime. https://github.com/pallets/jinja/issues/1194 Currently it is only specified as far as what the template docs describe, which is not sufficient. I would love help in this area from people with parser experience.

**Reply by cathaysia** [reply]
Thanks for you reply. I'm not a expert of parser, but I had wrote some parser already. maybe I should go back after tree-sitter-jinja2 more complete. maybe it can help #1194.

After read the discussion of #1194, it seems people tend to define jinja2 grammar by PEG grammar. But for my little experience on [pest](https://pest.rs/), PEG grammar cannot deal with the following case:

1. a sequence consist of `uint8_taaaaaa b`. It's clear that the type is `uint8_taaaaa`, and the variable is `b`.
2. PEG will parser the sequence to three token: `uint8_t`, `aaaaaa`, `b,` because uint8_t is a builtin type.
3. then the sequence parse failed.

I don't know where it is a implement problem. But maybe BNF grammar is better.

**Reply by davidism** [reply]
We don't define it by PEG, we're just interested in exploring it as the parser or grammar. I don't think your assessment of PEG is correct either, but that's not really important for this. Really it doesn't matter what grammar format we use, it can even be plain text sentences, as long as it's more specific and comprehensive than what's there now.

**Reply by cathaysia** [reply]
Yeah, I'm more inclined to think this is a problem with pest rather than PEG itself.
